File: backend/app/core/validation/rules.py
"""The single catalogue of input rules shared by the whole application.

Every user-controlled field in this system is described by exactly one
:class:`Rule`. The rule says what the field *is* — a person's name, a free-text
description, an email, a password, an identifier — and the limits below say how
long it may be. Nothing else in the backend should invent its own length cap or
its own email regex; before this module existed the email pattern was written
out twice in ``app/schemas/member.py`` alone and every schema picked its own
maximum length by hand.

The three clients of this catalogue are deliberately kept in step:

* ``backend/app/core/validation``  — the final authority (this package)
* ``desktop/core/validation``      — the same rules, for early feedback
* ``frontend/src/validation``      — the same rules again, in TypeScript

If you change a limit here, change it in the other two. The tests in
``backend/tests/test_validation_framework.py`` and
``desktop/tests/test_validation_framework.py`` assert the numbers agree.

Adding a new field? Pick the rule that matches its meaning. Only add a *new*
rule when a genuinely new kind of data arrives — not when an existing kind
needs a different length, which is what ``max_length`` overrides are for.
"""
from __future__ import annotations

import re
from enum import Enum


class Rule(str, Enum):
    """What kind of thing a field holds.

    The value doubles as the name used in error messages and in the desktop and
    frontend catalogues, so keep the spellings identical across the three
    layers.
    """

    #: A short human-readable label: a person, a project, a task, a title.
    NAME = "name"
    #: Multi-line prose written by a user: descriptions, reasons, comments.
    DESCRIPTION = "description"
    #: Single-line free text with no line breaks (a window title, a label).
    PLAIN_TEXT = "plain_text"
    #: An email address.
    EMAIL = "email"
    #: A secret. Never trimmed, never normalised, never pattern-checked.
    PASSWORD = "password"
    #: A whole number.
    INTEGER = "integer"
    #: A fractional number.
    DECIMAL = "decimal"
    #: A database identifier: a positive whole number.
    IDENTIFIER = "identifier"
    #: A UUID in canonical hyphenated form.
    UUID = "uuid"
    #: A calendar date.
    DATE = "date"
    #: A date and time.
    DATETIME = "datetime"
    #: A value drawn from a fixed set.
    ENUM = "enum"
    #: An absolute http(s) URL.
    URL = "url"
    #: A hostname such as ``docs.example.com``.
    DOMAIN = "domain"
    #: A user's search term. Held to plain text and escaped before it reaches
    #: a SQL ``LIKE``.
    SEARCH = "search"
    #: An opaque client-generated idempotency key.
    IDEMPOTENCY_KEY = "idempotency_key"


# ---------------------------------------------------------------------------
# Length limits
# ---------------------------------------------------------------------------
# These are the numbers the whole system agrees on. They exist so that a field
# cannot be used to store an arbitrary payload, and so that a legitimate value
# is never truncated. Where a column is narrower than the rule, the schema
# passes an explicit override rather than editing the shared number.

#: Longest name we accept. Matches the 150-char columns used for member,
#: project and task names.
NAME_MAX_LENGTH = 150
#: Names must contain at least one non-whitespace character.
NAME_MIN_LENGTH = 1

#: Long enough for a detailed hand-off note, short enough to bound a row.
DESCRIPTION_MAX_LENGTH = 5000

#: One line of free text. Window titles are the widest real user of this.
PLAIN_TEXT_MAX_LENGTH = 255

#: The maximum length of an email address, per RFC 5321.
EMAIL_MAX_LENGTH = 254

#: Password policy. The minimum is a floor on guessing cost; the maximum only
#: stops a megabyte of text reaching bcrypt. Neither bound alters the secret.
PASSWORD_MIN_LENGTH = 8
PASSWORD_MAX_LENGTH = 128

#: Search boxes. Short, because a search term is a filter and not a document.
SEARCH_MAX_LENGTH = 100

#: Absolute URLs, and the hostnames inside them.
URL_MAX_LENGTH = 2048
DOMAIN_MAX_LENGTH = 255

#: Client-generated idempotency keys, matching the ``client_event_id`` columns.
IDEMPOTENCY_KEY_MAX_LENGTH = 255

#: Identifiers are positive. Zero and negatives are always a bug or an attack,
#: never a real primary key. The ceiling is a signed 64-bit maximum, which is
#: the widest key any of our tables can hold.
IDENTIFIER_MIN = 1
IDENTIFIER_MAX = 9_223_372_036_854_775_807

#: The largest number of items a repeatable query parameter may carry, so a
#: caller cannot push thousands of ids into a single ``IN (...)`` clause.
MAX_LIST_PARAM_ITEMS = 200


# ---------------------------------------------------------------------------
# Patterns
# ---------------------------------------------------------------------------

#: Email. Deliberately the same shape the member schema already used, so this
#: refactor does not silently start rejecting addresses that used to work.
#: Full RFC 5322 is not worth implementing: it accepts addresses no provider
#: issues, and the real proof an address exists is a delivered message.
EMAIL_PATTERN = re.compile(r"^[^\s@]+@[^\s@]+\.[^\s@]+$")

#: Canonical hyphenated UUID, any version.
UUID_PATTERN = re.compile(
    r"^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-"
    r"[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$"
)

#: A hostname: dot-separated labels, no scheme, no path, no userinfo.
DOMAIN_PATTERN = re.compile(
    r"^(?=.{1,253}$)(?!-)[A-Za-z0-9-]{1,63}(?<!-)"
    r"(?:\.(?!-)[A-Za-z0-9-]{1,63}(?<!-))*$"
)

#: ``YYYY-MM-DD``. Range correctness is left to ``datetime.date.fromisoformat``.
DATE_PATTERN = re.compile(r"^\d{4}-\d{2}-\d{2}$")

#: Idempotency keys are opaque, so only the alphabet is constrained — enough to
#: keep them printable and safe to log.
IDEMPOTENCY_KEY_PATTERN = re.compile(r"^[A-Za-z0-9._:-]{1,255}$")

#: The schemes a stored URL may use. ``javascript:`` and ``data:`` are the two
#: that turn a stored link into script execution in whatever renders it.
ALLOWED_URL_SCHEMES = ("http", "https")


# ---------------------------------------------------------------------------
# Structured-content detection
# ---------------------------------------------------------------------------
# These are the *secondary* defence. The primary defence is that every field
# above has an expected type, a length and a format. What follows only rejects
# content that is structurally wrong for a plain-text field — markup, script,
# a whole JSON document, a whole XML document — and it is written to be
# specific so that ordinary prose survives.
#
# Deliberate trade-off, so that a future reader does not think it accidental:
# a sentence like "wrap it in a <div>" is rejected from a description, because
# the pattern cannot tell that apart from injected markup. Comparisons such as
# "5 < 10" and arrows such as "a -> b" are unaffected, because a tag requires
# "<" to be followed immediately by a letter or a slash.

#: An HTML/XML tag: ``<p>``, ``</div>``, ``<img src=x>``, ``<br/>``.
HTML_TAG_PATTERN = re.compile(r"</?[A-Za-z][A-Za-z0-9-]*(\s[^<>]*)?/?>")

#: An HTML entity that could reconstitute a tag after decoding, e.g. ``&lt;``,
#: ``&#60;``, ``&#x3c;``.
ENCODED_MARKUP_PATTERN = re.compile(
    r"&(?:lt|gt|#0*(?:60|62)|#[xX]0*3[cCeE]);"
)

#: A ``javascript:``/``vbscript:``/``data:`` URL, wherever it appears.
SCRIPT_URI_PATTERN = re.compile(
    r"\b(?:javascript|vbscript|data)\s*:", re.IGNORECASE
)

#: An inline event handler such as ``onerror=`` or ``onclick =``.
EVENT_HANDLER_PATTERN = re.compile(r"\bon[a-z]{3,20}\s*=", re.IGNORECASE)

#: A template-injection wrapper: ``${...}``, ``{{...}}``, ``<%...%>``.
TEMPLATE_EXPRESSION_PATTERN = re.compile(r"\$\{[^}]*\}|\{\{[^}]*\}\}|<%.*?%>", re.DOTALL)

#: An XML prolog or a CDATA section.
XML_PROLOG_PATTERN = re.compile(r"<\?xml\b|<!\[CDATA\[|<!DOCTYPE\b", re.IGNORECASE)

#: A NUL or other C0 control character. Tab, newline and carriage return are
#: excluded here and handled by the per-rule newline policy instead.
CONTROL_CHARACTER_PATTERN = re.compile(r"[\x00-\x08\x0b\x0c\x0e-\x1f\x7f]")


__all__ = [
    "Rule",
    "NAME_MAX_LENGTH",
    "NAME_MIN_LENGTH",
    "DESCRIPTION_MAX_LENGTH",
    "PLAIN_TEXT_MAX_LENGTH",
    "EMAIL_MAX_LENGTH",
    "PASSWORD_MIN_LENGTH",
    "PASSWORD_MAX_LENGTH",
    "SEARCH_MAX_LENGTH",
    "URL_MAX_LENGTH",
    "DOMAIN_MAX_LENGTH",
    "IDEMPOTENCY_KEY_MAX_LENGTH",
    "IDENTIFIER_MIN",
    "IDENTIFIER_MAX",
    "MAX_LIST_PARAM_ITEMS",
    "EMAIL_PATTERN",
    "UUID_PATTERN",
    "DOMAIN_PATTERN",
    "DATE_PATTERN",
    "IDEMPOTENCY_KEY_PATTERN",
    "ALLOWED_URL_SCHEMES",
    "HTML_TAG_PATTERN",
    "ENCODED_MARKUP_PATTERN",
    "SCRIPT_URI_PATTERN",
    "EVENT_HANDLER_PATTERN",
    "TEMPLATE_EXPRESSION_PATTERN",
    "XML_PROLOG_PATTERN",
    "CONTROL_CHARACTER_PATTERN",
]

File: frontend/src/validation/rules.ts
/**
 * The single catalogue of input rules for the web client.
 *
 * This file is a deliberate mirror of `backend/app/core/validation/rules.py`.
 * Every constant and every pattern below is the same value the backend
 * enforces, so that the browser can give immediate, accurate feedback and the
 * user is never told a value is fine only for the API to reject it (or the
 * reverse). The backend remains the final authority — this is early feedback,
 * not a replacement for it.
 *
 * **If you change a limit here, change it there too.** The three catalogues
 * (`backend/app/core/validation`, `desktop/core/validation`,
 * `frontend/src/validation`) are kept in step on purpose.
 */

/**
 * What kind of thing a field holds. The string values match the backend
 * `Rule` enum exactly, so a rule name can be read straight off either side.
 */
export type Rule =
  /** A short human-readable label: a person, a project, a task, a title. */
  | 'name'
  /** Multi-line prose written by a user: descriptions, reasons, comments. */
  | 'description'
  /** Single-line free text with no line breaks (a window title, a label). */
  | 'plain_text'
  /** An email address. */
  | 'email'
  /** A secret. Never trimmed, never normalised, never pattern-checked. */
  | 'password'
  /** A whole number. */
  | 'integer'
  /** A fractional number. */
  | 'decimal'
  /** A database identifier: a positive whole number. */
  | 'identifier'
  /** A UUID in canonical hyphenated form. */
  | 'uuid'
  /** A calendar date. */
  | 'date'
  /** A date and time. */
  | 'datetime'
  /** A value drawn from a fixed set. */
  | 'enum'
  /** An absolute http(s) URL. */
  | 'url'
  /** A hostname such as `docs.example.com`. */
  | 'domain'
  /** A user's search term. */
  | 'search'
  /** An opaque client-generated idempotency key. */
  | 'idempotency_key';

// ---------------------------------------------------------------------------
// Length limits — identical to the backend's
// ---------------------------------------------------------------------------

/** Longest name we accept. Matches the 150-char name columns. */
export const NAME_MAX_LENGTH = 150;
/** Names must contain at least one non-whitespace character. */
export const NAME_MIN_LENGTH = 1;

/** Long enough for a detailed hand-off note, short enough to bound a row. */
export const DESCRIPTION_MAX_LENGTH = 5000;

/** One line of free text. */
export const PLAIN_TEXT_MAX_LENGTH = 255;

/** The maximum length of an email address, per RFC 5321. */
export const EMAIL_MAX_LENGTH = 254;

/**
 * Password policy. The minimum is a floor on guessing cost; the maximum only
 * stops a megabyte of text reaching the hasher. Neither bound alters the secret.
 */
export const PASSWORD_MIN_LENGTH = 8;
export const PASSWORD_MAX_LENGTH = 128;

/** Search boxes. Short, because a search term is a filter and not a document. */
export const SEARCH_MAX_LENGTH = 100;

/** Absolute URLs, and the hostnames inside them. */
export const URL_MAX_LENGTH = 2048;
export const DOMAIN_MAX_LENGTH = 255;

/** Client-generated idempotency keys. */
export const IDEMPOTENCY_KEY_MAX_LENGTH = 255;

/**
 * Identifiers are positive. Zero and negatives are always a bug or an attack,
 * never a real primary key.
 */
export const IDENTIFIER_MIN = 1;

/**
 * The ceiling, which is the one number here that deliberately does *not* match
 * the backend's.
 *
 * The backend's `IDENTIFIER_MAX` is the signed 64-bit maximum,
 * 9_223_372_036_854_775_807. A JavaScript number cannot hold that value: it
 * rounds to 9223372036854775808, so writing it here would be a literal whose
 * runtime value is not the number written. More to the point, any id above
 * `Number.MAX_SAFE_INTEGER` has already lost precision by the time it reaches
 * this check, so accepting it would be accepting a value we can no longer
 * compare correctly.
 *
 * This bound is therefore *narrower* than the backend's, which is the safe
 * direction: everything the browser accepts, the API also accepts. Real keys
 * in this system are nowhere near either number.
 */
export const IDENTIFIER_MAX = Number.MAX_SAFE_INTEGER;

/** The largest number of items a repeatable query parameter may carry. */
export const MAX_LIST_PARAM_ITEMS = 200;

// ---------------------------------------------------------------------------
// Patterns
// ---------------------------------------------------------------------------

/**
 * Email. The same shape the backend uses. Full RFC 5322 is not worth
 * implementing: it accepts addresses no provider issues, and the real proof an
 * address exists is a delivered message.
 */
export const EMAIL_PATTERN = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;

/** Canonical hyphenated UUID, any version. */
export const UUID_PATTERN =
  /^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/;

/** A hostname: dot-separated labels, no scheme, no path, no userinfo. */
export const DOMAIN_PATTERN =
  /^(?=.{1,253}$)(?!-)[A-Za-z0-9-]{1,63}(?<!-)(?:\.(?!-)[A-Za-z0-9-]{1,63}(?<!-))*$/;

/** `YYYY-MM-DD`. Range correctness is checked separately. */
export const DATE_PATTERN = /^\d{4}-\d{2}-\d{2}$/;

/** Idempotency keys are opaque, so only the alphabet is constrained. */
export const IDEMPOTENCY_KEY_PATTERN = /^[A-Za-z0-9._:-]{1,255}$/;

/**
 * The schemes a stored URL may use. `javascript:` and `data:` are the two that
 * turn a stored link into script execution in whatever renders it.
 */
export const ALLOWED_URL_SCHEMES: readonly string[] = ['http', 'https'];

// ---------------------------------------------------------------------------
// Structured-content detection
// ---------------------------------------------------------------------------
// These are the *secondary* defence. The primary defence is that every field
// has an expected type, a length and a format. What follows only rejects
// content that is structurally wrong for a plain-text field — markup, script,
// a whole JSON document, a whole XML document — and it is written to be
// specific so that ordinary prose survives.
//
// Deliberate trade-off, so a future reader does not think it accidental: a
// sentence like "wrap it in a <div>" is rejected from a description, because
// the pattern cannot tell that apart from injected markup. Comparisons such as
// "5 < 10" and arrows such as "a -> b" are unaffected, because a tag requires
// "<" to be followed immediately by a letter or a slash.

/** An HTML/XML tag: `<p>`, `</div>`, `<img src=x>`, `<br/>`. */
export const HTML_TAG_PATTERN = /<\/?[A-Za-z][A-Za-z0-9-]*(\s[^<>]*)?\/?>/;

/** An HTML entity that could reconstitute a tag after decoding. */
export const ENCODED_MARKUP_PATTERN = /&(?:lt|gt|#0*(?:60|62)|#[xX]0*3[cCeE]);/;

/** A `javascript:`/`vbscript:`/`data:` URL, wherever it appears. */
export const SCRIPT_URI_PATTERN = /\b(?:javascript|vbscript|data)\s*:/i;

/** An inline event handler such as `onerror=` or `onclick =`. */
export const EVENT_HANDLER_PATTERN = /\bon[a-z]{3,20}\s*=/i;

/** A template-injection wrapper: `${...}`, `{{...}}`, `<%...%>`. */
export const TEMPLATE_EXPRESSION_PATTERN = /\$\{[^}]*\}|\{\{[^}]*\}\}|<%[\s\S]*?%>/;

/** An XML prolog or a CDATA section. */
export const XML_PROLOG_PATTERN = /<\?xml\b|<!\[CDATA\[|<!DOCTYPE\b/i;

/**
 * A NUL or other C0 control character. Tab, newline and carriage return are
 * excluded here and handled by the per-rule newline policy instead.
 */
// oxlint-disable-next-line no-control-regex
export const CONTROL_CHARACTER_PATTERN = /[\x00-\x08\x0b\x0c\x0e-\x1f\x7f]/;

/**
 * The maximum length each rule enforces, for driving `maxLength` on an input.
 * Rules with no natural character cap are absent.
 */
export const RULE_MAX_LENGTH: Partial<Record<Rule, number>> = {
  name: NAME_MAX_LENGTH,
  description: DESCRIPTION_MAX_LENGTH,
  plain_text: PLAIN_TEXT_MAX_LENGTH,
  email: EMAIL_MAX_LENGTH,
  password: PASSWORD_MAX_LENGTH,
  search: SEARCH_MAX_LENGTH,
  url: URL_MAX_LENGTH,
  domain: DOMAIN_MAX_LENGTH,
  idempotency_key: IDEMPOTENCY_KEY_MAX_LENGTH,
};
